Add empty folder indicator

PVince81 opened the pull request:
## Description
Add empty folder indicator

## Related Issue
Fixes https://github.com/owncloud/phoenix/issues/1910

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- test environment:
- test case 1:
- test case 2:
- ...

## Screenshots (if appropriate):
<img width="959" alt="image" src="https://user-images.githubusercontent.com/277525/74037911-8eb48d80-49bf-11ea-8d1c-33d4c668b1c3.png">


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [x] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
- [x] improve design, the colors don't match. Bigger size would be nice.
- [x] also process the other lists like "Favorites", "Shared with me", etc
- [x] changelog

PVince81 commented:
I've added acceptance tests and also tweaked the code that retrieves the list to avoid another potentially waiting code when waiting is not needed => faster results

PVince81 commented:
@individual-it I've pinged you as the tests part is less obvious. I've twisted them a bit to include the "empty message" detection inside the test step that checks if the resource list is empty.

The reason for that is that I wanted to avoid UI-specific language in the Gherkin step like "a message must appear" or something, so I thought having it implied but still checked through assertions would be the way to go. Let me know what you think.

LukasHirt commented:
> the colors don't match

If you leave out the variation in the icon it will get "system" (default) variation which has the same colour as meta text. Though I'd see this as a workaround only. But until we bring real oC colours into ODS I'd say it's good enough.

LukasHirt commented on apps/files/src/components/AllFilesList.vue:
```diff
@@ -37,6 +37,9 @@
             </div>
           </template>
+          <template #emptyMessage>
+            <div class="uk-text-center"><oc-icon :name="$_emptyFolderIcon" type="div" size="large" variation="passive" /><div class="uk-text-muted" v-text="$_emptyFolderText" /></div>
```
Separate this into own lines pls, it's too long.

LukasHirt commented on apps/files/src/components/Collaborators/SharedFilesList.vue:
```diff
@@ -45,6 +45,9 @@
       <div class="uk-text-meta uk-text-nowrap uk-width-small" v-text="formDateFromNow(item.shareTime)" />
     </template>
+    <template #emptyMessage>
+      <div class="uk-text-center"><oc-icon name="share" type="div" size="large" variation="passive" /><div class="uk-text-muted" v-text="$_emptyFolderText" /></div>
```
Separate this into own lines pls, it's too long.

LukasHirt commented on apps/files/src/components/Trashbin.vue:
```diff
@@ -30,6 +30,9 @@
           </div>
         </template>
+        <template #emptyMessage>
+          <div class="uk-text-center"><oc-icon name="delete" type="div" size="large" variation="passive" /><div class="uk-text-muted" v-translate>No deleted files</div></div>
```
Separate this into own lines pls, it's too long.

LukasHirt commented on apps/files/src/components/AllFilesList.vue:
```diff
@@ -99,6 +102,26 @@ export default {
+
+    $_emptyFolderText () {
+      if (this.$_isFavoritesList) {
+        return this.$gettext('No favorites defined')
```
Maybe too much technical message? What about something like "No files/folders marked as favorite"?

kulmann commented on apps/files/src/components/FilesAppBar.vue:
```diff
@@ -50,7 +50,7 @@
               <translate>Restore selected</translate>
             </oc-button>
-            <oc-button id="delete-selected-btn" icon="delete" @click="selectedFiles.length < 1 ? $_ocTrashbin_empty() : $_ocTrashbin_deleteSelected()">
+            <oc-button id="delete-selected-btn" icon="delete" :disabled="files.length === 0" @click="selectedFiles.length < 1 ? $_ocTrashbin_empty() : $_ocTrashbin_deleteSelected()">
```
Good catch to disabled the button! However, you can now remove the part from the click handler, that deals with an empty selection.

PVince81 replied:
empty selection doesn't mean no files
if you click the button with no files selected, it's the same as selecting all

PVince81 commented:
I've addressed all comments, except for the design part.

- [ ] We need to update ODS to get the bigger icon sizes: https://github.com/owncloud/owncloud-design-system/pull/632
- [ ] Set size to xlarge and make sure the font size also matches

individual-it commented:
In the spirit of BDD and the discussions we had in the product repo about how to write steps I would suggest to explicitly have a step that says that there should be the message. I would argue that there is a new behaviour in the application, but that is not reflected in the feature files.
@HanaGemela @phil-davis how do you feel about that?

individual-it commented on tests/acceptance/stepDefinitions/filesContext.js:
```diff
@@ -320,10 +320,16 @@ When('the user unmarks the favorited file/folder {string} using the webUI', func
+
+  const allRowsResult = await client.page.FilesPageElement.filesList().allFileRows()
+
+  assert.ok(emptyMessageVisible, 'Empty message must be visible')
```
is the message empty, or is the message about the folder being empty? ;-)

PVince81 replied:
the message is about "folder is empty", see screen shot above.

I suspected that this wording will cause confusion, I'll adjust it

PVince81 replied:
renamed to "noContentMessage"

PVince81 commented:
> In the spirit of BDD and the discussions we had in the product repo about how to write steps I would suggest to explicitly have a step that says that there should be the message. I would argue that there is a new behaviour in the application, but that is not reflected in the feature files.
> @HanaGemela @phil-davis how do you feel about that?

I was also a bit confused here and went for the idea that the "empty folder" message is an implementation/design detail (which could be considered obvious in any web app)) not related to the requirements. I guess if we say that it's an explicit product requirement to display a message, then we need it as a Gherkin step.

kulmann commented on apps/files/src/components/AllFilesList.vue:
```diff
@@ -118,9 +118,17 @@ export default {
+      if (this.$_isFavoritesList) {
+        return this.$gettext('You can mark some by clicking on the star icon.')
+      } else {
+        return this.$gettext('Drag files and folder here to upload, or browse files from your computer.') // TODO: link
```
- `folder` missing an s (plural)
- maybe replace `computer` with `device`, as we're also targeting mobile

kulmann commented on apps/files/src/components/AllFilesList.vue:
```diff
@@ -118,9 +118,17 @@ export default {
+
+    $_emptyFolderSubText () {
+      if (this.$_isFavoritesList) {
+        return this.$gettext('You can mark some by clicking on the star icon.')
```
... `star icon in the file list`? As a new user I might not know where to find the star icon.

PVince81 commented:
Fixed all "no content" messages.
Renamed EmptyMessage component to NoContentMessage to avoid confusion.
Separate styling for message and call to action.
Adjusted screenshot in original post.

@kulmann please recheck.

kulmann commented on apps/files/src/components/Collaborators/SharedFilesList.vue:
```diff
@@ -45,6 +45,14 @@
+      <no-content-message icon="share">
+        <template #message>
+          <span v-if="$_isSharedWithMe"><translate>You are currently not collaborating on other people's resources.</translate></span>
+          <span v-if="!$_isSharedWithMe"><translate>You are currently not collaborating on any of your resources with other people.</translate></span>
```
`v-else`?

PVince81 commented:
For some reason I intuitively tend to avoid `v-else` as it makes the element depend on the previous one, which means I can't easily move around the element with `v-else`. in some cases, like the RecycleScroller, one needs to scroll a lot to find the original condition.

Anyway, I think we can agree to use `v-else` still. I've adjusted accordingly.

LukasHirt commented on apps/files/src/components/AllFilesList.vue:
```diff
@@ -37,6 +37,16 @@
           </template>
+          <template #noContentMessage>
+            <no-content-message v-if="!$_isFavoritesList" icon="folder">
+              <template #message><translate>There are no resources in this folder.</translate></template>
```
Why not using the slot directly on `translate` tag?

PVince81 replied:
Do you mean putting v-translate on the template tag ? I wasn't sure whether this works.
I guess I could test it with a dummy temporary local translation

LukasHirt replied:
Not sure if that works. I thought more about `<translate v-slot:message>`

LukasHirt replied:
Ah, now I see this was deprecated in 2.6.0. My bad...

LukasHirt commented on apps/files/src/components/Collaborators/SharedFilesList.vue:
```diff
@@ -45,6 +45,14 @@
+    <template #noContentMessage>
+      <no-content-message icon="share">
+        <template #message>
+          <span v-if="$_isSharedWithMe"><translate>You are currently not collaborating on other people's resources.</translate></span>
```
`translate` tags default to span. Could we then either use `v-translate` directly on the span or dropping the span and moving the if statement to the translate tag? I'd prefer the former.

PVince81 commented:
@LukasHirt I've addressed your comments and moved the `v-translate` on the previous tag.

also fixed the tests

PVince81 commented:
I did a few tests with translations:
- `<template #message v-translate>` the text stays in english
- `<translate v-slot:message>` kills the text completely
- `<span v-translate>` works correctly

PVince81 commented:
@LukasHirt I went with `<span v-translate>` now to avoid span duplication.

LukasHirt commented:
@PVince81 Thanks for those tests! I saw that the `v-translate` is now in some parts on `<template>`. If I got your test right it's not translating, right? Maybe just a leftover from tests then?

PVince81 commented:
@LukasHirt forgot pushing, please rechech

PVince81 commented:
I messed up the test sentences while doing manual search and replace... Fixed now.

PVince81 commented:
Yeah, the tests fail because they use an existing skeleton folder which contains the file `.gitkeep` :palm_tree: :face_with_head_bandage:

PVince81 commented:
Seems the "no content" check is not compatible with the files drop page, so I've added an URL check and disabled the specific check for the message presence when on the files-drop page.

LukasHirt commented on apps/files/src/components/FileList.vue:
```diff
@@ -2,7 +2,7 @@
   <div :id="id" class="uk-height-1-1 uk-position-relative" @click="hideRowActionsDropdown">
     <div class="uk-flex uk-flex-column uk-height-1-1">
-      <oc-grid gutter="small" flex id="files-table-header" class="uk-padding-small">
+      <oc-grid gutter="small" flex id="files-table-header" class="uk-padding-small" v-if="fileData.length > 0" >
```
Add key pls to `v-if`/`v-else`
